fml/task_runner.h

```diff
@@ -62,6 +66,11 @@ class TaskRunner : public fml::RefCountedThreadSafe<TaskRunner>,
   static void RunNowOrPostTask(fml::RefPtr<fml::TaskRunner> runner,
                                const fml::closure& task);
 
+  /// Similar to |RunNowOrPostTask|, but it needs to synchronize with thread
+  /// merging operation.
+  static void RunNowOrPostSyncTask(fml::RefPtr<fml::TaskRunner> runner,
+                                   const fml::closure& task);
+
```

Comment:
I would avoid adding this new method - someone is likely to use the wrong one not realizing they were supposed to use this one because of thread merging.

Reply:
Here add a lock to work around data race caused by the raster thread performing thread merging in the process of the platform thread post sync task to the raster thread. 

Considering that most scenarios where `RunNowOrPostTask` is called do not require the lock and should not suffer a performance penalty, finally I choose to add this new method.  It would be best if there was a workaround that could take care of both.

Reply:
I agree with @dnfield , this method can be misused. Actually this was discussed before.
https://github.com/flutter/engine/pull/28159#discussion_r695006096

Reply:
> I agree with @dnfield , this method can be misused. Actually this was discussed before. [#28159 (comment)](https://github.com/flutter/engine/pull/28159#discussion_r695006096)

It seems a good choice.

Reply:
```c++
void TaskRunner::RunNowOrPostTaskSync(fml::RefPtr<fml::TaskRunner> runner,
                                      const fml::closure& task) {
  FML_DCHECK(runner);
  std::scoped_lock lock(g_thread_merging_lock);
  if (runner->RunsTasksOnCurrentThread()) {
    task();
  } else {
    fml::AutoResetWaitableEvent latch;
    runner->PostTask([&] {
      task();
      latch.Signal();
    });
    latch.Wait();
  }
}
```

@dnfield @ColdPaleLight  

What about this sync method? It encapsulates the latch synchronization operation into the function. It may mitigate the probability of misuse.  After all, it is quite different from `RunNowOrPostTask`. :)

Reply:
I think the bigger problem is how do I know when I have to use this method, and why should I ever choose `RunNowOrPostTask` over this one? 

/cc @chinmaygarde